File: 002-dynamic-programming/004-coin-combinations-2.py

n, x = [int(s) for s in input().split()]
c = [int(s) for s in input().split()]
MOD = 1000000007
dp = [[0 for i in range(x + 1)] for j in range(n + 1)]

for i in range(n):
    dp[i][0] = 1

for i in range(n - 1, -1, -1):
    for j in range(1, x + 1):
        skipping = dp[i + 1][j]
        picking = 0
        if c[i] <= j:
            picking = dp[i][j - c[i]]
        dp[i][j] = (skipping + picking) % MOD
print(dp[0][x])

File: 002-dynamic-programming/004-coin-combinations-2.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n, x;
    cin >> n >> x;
    vector<int> c(n);
    for (int i = 0; i < n; ++i) {
        cin >> c[i];
    }
    vector<vector<int>> dp(n + 1, vector<int>(x + 1));
    int MOD = 1000000007;

    // base case
    for(int i = 0; i < n; i++){
        dp[i][0] = 1;
    }

    for(int i = n - 1; i >= 0; i--){
        for(int j = 1; j <= x; j++){
            int skipping = dp[i + 1][j];
            int picking = 0;
            if(c[i] <= j){
                picking = dp[i][j - c[i]];
            }
            dp[i][j] = (skipping + picking) % MOD;
        }
    }

    cout<<dp[0][x]<<endl;
    return 0;
}